warehouse/upload.go

```diff
@@ -431,6 +443,101 @@ func (job *UploadJobT) run() (err error) {
 	return nil
 }
 
+// TableUploadStatusT captures the status of each table upload along with its parent upload_job's info like destionation_id and namespace
+type TableUploadStatusT struct {
+	uploadID      int64
+	destinationID string
+	namespace     string
+	tableName     string
+	status        string
+}
+
+func (job *UploadJobT) fetchPendingUploadTableStatus() []*TableUploadStatusT {
+	//TODO: Get only tables' status of current uploadJob
+	sqlStatement := fmt.Sprintf(`
+		SELECT
+			%[1]s.id,
+			%[1]s.destination_id,
+			%[1]s.namespace,
+			%[2]s.table_name,
+			%[2]s.status
+		FROM
+			%[1]s INNER JOIN %[2]s
+		ON
+			%[1]s.id = %[2]s.wh_upload_id
+		WHERE
+			%[1]s.id <= '%[3]d'
+			AND %[1]s.destination_id = '%[4]s'
+			AND %[1]s.namespace = '%[5]s'
+			AND %[1]s.status != '%[6]s'
+			AND %[1]s.status != '%[7]s' 
+			AND %[1]s.table_name in (SELECT table_name FROM %[2]s WHERE %[2]s.wh_upload_id = '%[3]d')
+		ORDER BY
+			%[1]s.id ASC`,
+		warehouseutils.WarehouseUploadsTable,
+		warehouseutils.WarehouseTableUploadsTable,
+		job.upload.ID,
+		job.upload.DestinationID,
+		job.upload.Namespace,
+		ExportedData,
+		Aborted)
+	rows, err := job.dbHandle.Query(sqlStatement)
+	if err != nil && err != sql.ErrNoRows {
+		panic(err)
+	}
+	defer rows.Close()
+
+	tableUploadStatuses := make([]*TableUploadStatusT, 0)
+
+	for rows.Next() {
+		var tableUploadStatus TableUploadStatusT
+		err := rows.Scan(
+			&tableUploadStatus.uploadID,
+			&tableUploadStatus.destinationID,
+			&tableUploadStatus.namespace,
+			&tableUploadStatus.tableName,
+			&tableUploadStatus.status,
+		)
+		if err != nil {
+			panic(err)
+		}
+		tableUploadStatuses = append(tableUploadStatuses, &tableUploadStatus)
+	}
+
+	return tableUploadStatuses
+}
+
+func getTableUploadStatusMap(tableUploadStatuses []*TableUploadStatusT) map[int64]map[string]string {
+	tableUploadStatus := make(map[int64]map[string]string)
+	for _, tUploadStatus := range tableUploadStatuses {
+		if _, ok := tableUploadStatus[tUploadStatus.uploadID]; !ok {
+			tableUploadStatus[tUploadStatus.uploadID] = make(map[string]string)
+		}
+		tableUploadStatus[tUploadStatus.uploadID][tUploadStatus.tableName] = tUploadStatus.status
+	}
+	return tableUploadStatus
+}
+
+func (job *UploadJobT) getTablesToSkip() []string {
+	tableUploadStatuses := job.fetchPendingUploadTableStatus()
+	tableUploadStatus := getTableUploadStatusMap(tableUploadStatuses)
+	skipTableMap := make(map[string]bool)
+	for uploadID, tableStatusMap := range tableUploadStatus {
+		for tableName, status := range tableStatusMap {
+			if (uploadID < job.upload.ID && status == TableUploadExportingFailed) || //Previous upload and table upload failed
+				(uploadID == job.upload.ID && status == TableUploadExported) { //Current upload and table upload succeeded
```

Comment:
we also check if current table has already been exported using func shouldTableBeLoaded in upload.go. Should we remove/edit that also?

Reply:
Yea. I couldn't move this logic there. Because we don't have enough context to find out if the tables of previous jobs are loaded or not. Also I considered removing it there, but didn't want to disturb it.

Reply:
Actually shouldTableBeLoaded is used in loadUserTables and loadIdentityTables where our table skip logic is not there. So instead of changing shouldTableBeLoaded, I replaced it's usage in loadTablesExcept with only hasLoadFiles. Take a look once.